### ISSUE TITLE
Touch screen trigger and point source

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#23]

**Comment by JuliusSweetland** [comment]
Moving to BACKLOG

**Comment by JuliusSweetland** [comment]
Resurrected

**Comment by JuliusSweetland** [comment]
* FrameReported event (WPF):
http://stackoverflow.com/questions/4825959/filtering-a-touch-framereported-iobservable-using-arbitrary-boolean-condition-th
https://msdn.microsoft.com/en-us/library/system.windows.input.touch.framereported(v=vs.110).aspx

* Window_Manipulation events (also WPF) - I think these are high level versions of the WM_MANIPULATION events:
https://msdn.microsoft.com/en-GB/library/ee649090(v=vs.100).aspx

* Alternatively override WinProc and listen for WM_POINTERXXX events:
http://stackoverflow.com/questions/23940143/recognize-touch-as-mousedown-event

* Also override WinProc and listen for WM_TOUCH events?

* GitHub project which handles some touch gestures: https://github.com/rprouse/WinTouch.NET

**Comment by alexandre-mbm** [comment]
What means "point source"?

**Comment by JuliusSweetland** [comment]
@alexandre-mbm A point source is any input device which can supply a list of points (x,y coordinates on the screen) to tell OptiKey where the user is interested. An eye tracker is a point source. So is a mouse. In this issue where the user touches the screen would be the point of interest, so a touch screen point source would emit points when the user was touching the screen, or moving their finger when in contact with the screen, but stop when they stopped touching the screen. This is an issue I've been wanting to work on for a long time! Let me know if you are interested in working on it.

**Comment by alexandre-mbm** [comment]
> This is an issue I've been wanting to work on for a long time! Let me know if you are interested in working on it.

No. Sorry. My father cannot enjoy a touch screen and I am with no time for focusless work.

**Comment by JuliusSweetland** [comment]
WM_TOUCH (windows 7 onward) limitations and why WM_POINTER (win 8 onwards) is better: http://the-witness.net/news/2012/10/wm_touch-is-totally-bananas/comment-page-1/

**Comment by JuliusSweetland** [comment]
This is how to hook events that are routed to a specific window handle, but not global events: https://stackoverflow.com/questions/23940143/recognize-touch-as-mousedown-event

**Comment by JuliusSweetland** [comment]
Discussion of hooking global pointer events, but no definite answer: https://stackoverflow.com/questions/14297322/how-to-detect-tapping-touch-input-globally-instead-of-mouse-clicking

**Comment by JuliusSweetland** [comment]
How to implement global hooks (using a C or C++ dll): https://stackoverflow.com/questions/17747345/callwndproc-example

**Comment by JuliusSweetland** [comment]
Discussion of global WM_TOUCH event hooking: https://social.microsoft.com/Forums/en-US/8812b649-c4f8-443b-af09-6f6866866833/how-to-catch-wmtouch-in-c-with-global-hook?forum=Offtopic

**Comment by JuliusSweetland** [comment]
Global hooks example with code: https://www.codeproject.com/Articles/18638/Using-Window-Messages-to-Implement-Global-System-H

**Comment by idanrichman** [comment]
Hi Julius,
I'm interested in adding touchscreen as a point source.
the reason for that is that the regular touchscreen as a mouse that windows inherently support isn't good enough for cases where the touch is coarse like when both the palm and the pointing finger are put on screen so it doesn't know where to position the mouse cursor (for people who lost their fine motor skills).
my initial thought was a simple rule of taking the most left-upper touch point (for right hands) or most right-upper (for left hands). I still see that when the whole palm pad is put on screen, it doesn't always recognize other touch points (touchpoint capacity is overloaded or something) but it might be good enough.

`System.Windows.Input` has a convenient Touch object that can get all touchpoints upon an event.

just for a quick testing purpose i took over the MousePositionSource.cs:
the problem is that `e.GetTouchPoints(null)` returns wrong coordinates - either screen or window or whatever, but not those that align with the rest of your code. the null should be replaced with the proper RelativeTo IInputElement. I believe you'll know how to solve this.

```
namespace JuliusSweetland.OptiKey.Observables.PointSources
{
    public class MousePositionSource : IPointSource
    {
        Point touchPt = new Point();
        private readonly TimeSpan pointTtl;

        private IObservable<Timestamped<PointAndKeyValue>> sequence;

        public MousePositionSource(TimeSpan pointTtl)
        {
            this.pointTtl = pointTtl;
            Touch.FrameReported += new TouchFrameEventHandler(Touch_FrameReported);
        }

        void Touch_FrameReported(object sender, TouchFrameEventArgs e)
        {
            touchPt.X = e.GetTouchPoints(null)[0].Position.X;
            touchPt.Y = e.GetTouchPoints(null)[0].Position.Y;
            foreach (TouchPoint _touchPoint in e.GetTouchPoints(null))
            {
                if ((_touchPoint.Position.X < touchPt.X) & (_touchPoint.Position.Y < touchPt.Y))
                {
                    touchPt.X = _touchPoint.Position.X;
                    touchPt.Y = _touchPoint.Position.Y;
                }
            }
        }

        public RunningStates State { get; set; }

        public Dictionary<Rect, KeyValue> PointToKeyValueMap { private get; set; }

        public IObservable<Timestamped<PointAndKeyValue>> Sequence
        {
            get
            {
                if (sequence == null)
                {
                    sequence = Observable
                        .Interval(Settings.Default.PointsMousePositionSampleInterval)
                        .Where(_ => State == RunningStates.Running)
                        //.Select(_ => new Point(System.Windows.Forms.Cursor.Position.X, System.Windows.Forms.Cursor.Position.Y))  // this is where it gets it's coordinates
                        .Select(_ => new Point(touchPt.X, touchPt.Y))  // this is where it gets it's coordinates
                        .Timestamp()
                        .PublishLivePointsOnly(pointTtl)
                        .Select(tp => new Timestamped<PointAndKeyValue>(tp.Value.ToPointAndKeyValue(PointToKeyValueMap), tp.Timestamp))
                        .Replay(1) //Buffer one value for every subscriber so there is always a 'most recent' point available
                        .RefCount();
                }

                return sequence;
            }
        }
    }
}
```

**Comment by JuliusSweetland** [comment]
Hi Idan,

I've looked at this method before and I'm not sure it will work, but we can
try. I believe you have to pass a control to that method to get the
relative position of the touch:

https
://docs.microsoft.com/en-us/dotnet/api/system.windows.input.touch.framereported?view=netframework-4.8
<https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.touch.framereported?view=netframework-4.8>

And

https://www.codeproject.com/Articles/692286/WPF-and-multi-touch

And lots of good overview into here:

https://stackoverflow.com/questions/26038294/how-can-i-get-the-touch-inputpositions-on-the-screen-on-the-windows-surface-pr

The first problem is trying to hack in a reference to a control. I would
recommend the whole OptiKey window. I'm not sure the window is created when
the point source is created, but you might be able to pass in something
which will be populated with the window later?

**Comment by kmcnaught** [comment]
@idanrichman just checking if you had made any progress here that could help us?

**Comment by idanrichman** [comment]
hi @kmcnaught, i'll need to check and find the source code, it was a while ago. 
> my initial thought was a simple rule of taking the most left-upper touch point (for right hands) or most right-upper (for left hands)
from what i do remember, i think this simple method worked quite nice. 

> I still see that when the whole palm pad is put on screen, it doesn't always recognize other touch points (touchpoint capacity is overloaded or something) but it might be good enough.
i think this was unsolved. some laptops (like microsoft's surface) has built-in mechanisms for to prevent palm touch, so i used that.

**Comment by idanrichman** [comment]
Hi,
sorry for the long delay.
anyhow, you can look at this commit
https://github.com/idanrichman/OptiKey-old/commit/b9ff664a32f7c1d8b1d5b59efe9485248124a342

it is based on quite an old version, but i think it will still hold.
it worked when i tried it on my laptop. touch was ok, and it stops "pressing" when finger is up.
it still doesn't solve when whole palm is on the screen, and can get sub-optimal results.
I have a piece of code I commented out, that tries to take the most upper-left (right-handed pointing finger) touch point. I don't remember why i commented it, and if it relates to the issue i raised before about aligning with correct screen coordinates. it still works without it.

hope this helps.

**Comment by kmcnaught** [comment]
Hi @idanrichman thank you for sharing your work. I will try to find some time to look at it soon - it would be great if we could get this working.

**Comment by kmcnaught** [comment]
@JuliusSweetland I'm looking at this. How were you imagining it interacting with trigger source? Would it only ever make sense to touch as [both point source and trigger], or would you envisage you could choose touch as point source and e.g. switch as trigger? Currently if you use left click as trigger with touch as point source there's some funky behaviour with repeating the previous key (so everything is triggered twice) but I'm sure that's easy enough to fix

**Comment by JuliusSweetland** [comment]
I did only imagine using touch as the trigger as well, which would also
allow you to swipe full words (in theory!)

**Comment by kmcnaught** [comment]
A few thoughts. 

I've adapted the proposed class to use events instead of polling, and use `GetPrimaryTouchPoint` instead of cursor position. This fixes an issue where the mouse and touch positions were slightly out of sync, so you got new touch events with stale cursor positions. 

This is the current implementation, for reference. It's on a slightly unstable branch but is probably usable if you wanted to try it. 
https://github.com/kmcnaught/OptiKey/blob/710ecc8940ccbe4c4330f90c9f06e4f610a31464/src/JuliusSweetland.OptiKey.Core/Observables/PointSources/TouchScreenPositionSource.cs

This works for simple key selection on my Surface tablet, and _almost_ works for swipe depending on settings.

A few thoughts/issues came up:

- The `TouchFrameEventHandler` events seem to only occur within the bounds of the Optikey app, so point selection elsewhere is not possible.
  - It looks like there may be ways round this, pasting links for future reading
  - https://stackoverflow.com/questions/56899214/wpf-multitouch-listen-to-system-wide-touchdown-events-in-windows-10
  - https://stackoverflow.com/questions/45473673/how-to-distinguish-touch-vs-mouse-event-from-setwindowshookex-in-c-sharp
  - Alternatively, we might consider whether we might decide that for point selection, falling back to mouse clicks is sufficient and appropriate, and we don't need touch here. 
  
- Swipe has the logically correct behave when in `stop signal = trigger pressed again` mode, but this isn't what we want. We want `stop signal = trigger released` mode, which is behaving weirdly and needs investigating. I tried comparing it to gamepad trigger input, which doesn't seem to respect `stop signal = trigger released` mode either, but in a less-bad way. I had naively not been aware of this distinction and should do a round of testing to check that the various event handling changes haven't broken something here.

- So far this implementation only works on a single touchscreen, no monitor. We should scope out what we want here. For targeting AAC use, this might be considered sufficient. As an advanced macro UI (as Anna has suggested) users might want to use it on their touch-enabled laptop while using an extended desktop on a monitor for example. I reckon I would use it like this! For this use-case we might also want to look at whether we can prevent Optikey from taking focus when touched, so that pressing a key doesn't move focus from the current app. 
 
Some boring things to note as TODOs:
- [ ] Update "ToExtendedString" for new point source
- [ ] Fix logic for encouraging / limiting trigger source choice depending on point source
- [ ] Prevent window dragging behaviour 
- [ ] Hide smoothing options when touch is selection as point source

**Comment by kmcnaught** [comment]
wow, past-me actually wrote some good notes here. 

I've been looking into the multikey selection behaviour here. The behaviour I'd been seeing was that the swipe gesture didn't get processed until you'd pressed the final key a second time (even though I'm using `"stop signal = trigger released"`. This was thoroughly confusing until I realised that the fact my code is event-based and not polling-driven means that the finalisation of the swipe gesture occurs on a touch "up" event but then there aren't any touch position events to get it picked up here:
```
var pointAndKeyValueSubscription = pointSource.Sequence
                    .Where(tp => tp.Value != null) //Filter out stale indicators
                    .Select(tp => new Timestamped<PointAndKeyValue>(tp.Value, tp.Timestamp))
                    .TakeWhile(tp => stopMultiKeySelectionTriggerSignal == null)
                    .ToList()
                    .Subscribe(points =>
                    {
                        observer.OnNext(points);
                        observer.OnCompleted();
                    });
                    ```
So we don't get past `TakeWhile...` until the next time we touch the screen. 

I think it's the case that all our point sources are either based on polling or are event-driven from a system (eye tracker engine) that runs on a regular clock, so gives a similar effect. In this case we don't get position events unless a finger is touching the screen (and therefore probably interacting with a "click" although maybe not in the case where someone wanted to use a different trigger). A few questions arise, some of which might be resolved for us when I explore the other limitations I noted above.

1 - Would it be better to just revert to polling? It's important we still use the touch information and not the cursor position since these were out of sync, but we could cache any state we need to and still end up with a regular stream

2 - or we could stay event driven, but recycle touch events with an `Observable.Interval`

3 - or should our architecture support input sources like this, which may have long gaps between events? I suspect there is various sequence logic that would need adjusting, possibly not just this situation. 

4 - ...or we could do the recycling trick on a higher level, so that any point source gets recycled at a certain interval. Though this might be inefficient if the point source happens to have a slightly longer interval (so we process double the events without needing to)

Thoughts?

I currently have an implementation for (2) which seems like a reasonable way forward right now, but also feels like something worth discussing. 

I am also still reflecting on how this touch implementation might be expected to work with other trigger sources. I can definitely picture a scenario where a user with limited motor control has a hand permanently resting on a tablet, and uses another input (such as a switch) to select the highlighted key. I am less convinced that using touch + dwell is a likely combination. That's not to say it shouldn't be supported, but there might be implications of the above approach which limits how dwell works (removing your touch wouldn't stop the countdown, as there would still be pseudo-touch events). I think this is probably okay.

**Comment by kmcnaught** [comment]
@JuliusSweetland also a separate discussion about MultiKey selection with touch input. If I put `MultiKeySelectionFixationMinDwellTime` down to `1ms` then swiping with touch works really pretty well. It has a slight bias towards longer words, as short words often have lots of intermediate letters, but don't pin you down much. Whereas long words work really well. 

I would argue that a minimum dwell time for touch input doesn't really make sense at all, and I'm sure users won't think to change it. More generally I'd argue that the 'dwell' time should just be used as a weighting but that's another conversation ;)

Should we just disable minimum dwell time for touch input? How about mouse input?

**Comment by JuliusSweetland** [comment]
Good news that it works fairly well already. I wonder if all that is needed
is a change to the auto space logic; perhaps if the multi key select only
outputs a single letter then it can skip the auto space? Is that possible
with the way it is written?

**Comment by kmcnaught** [comment]
@JuliusSweetland I need some input on preferred UX for changing settings together. I am of the opinion that 99% of the time, if someone wants Touch as a PointSource they will also want:
- touch up/down as a TriggerSource; and 
- MultiKeySelectionFixationMinDwellTime around 5ms (rather than default 250ms)

I think it's really important that we help users with this. Navigating a different source + key trigger + point trigger is a challenge I've seen in many users. The `MultiKeySelectionFixationMinDwellTime` is obscure and most people won't discover it, but it's hopeless at the default setting. 

I also don't think these settings should be completely overridden - it might be appropriate for someone to use touch as a point source (maybe with a hand permanently resting on the keyboard) and a separate switch as a trigger (that they can use in their own time). 

So what is your preferred way to coerce people's choices? We do have an existing pattern for coercing language settings for Urdu and Persian - this relies on a messagebox + optional correction at the point that the settings are saved (i.e. closing the management console). Is this the best approach here, or can we do something a little more immediate?

**Comment by JuliusSweetland** [comment]
I think we should use the same pattern at Urdu etc. Message box explaining
the changes and then coerce for them. We could do this in bother directions
so if they select touch and then change it to something else we set the
default dwell time back to 250ms.

**Comment by kmcnaught** [comment]
good shout, I'll do it in both directions, and also make sure they are not left with a touch trigger and a non-touch source

**Comment by JuliusSweetland** [comment]
Awesome, thanks!

**Comment by kmcnaught** [comment]
I've done the settings-coercion and tidied up a lot of the loose ends, and thought it was all coming together and then remembered that I haven't solved the issue of not being able to get touch events outside the app bounds. 

## what do we even want here?

From a user's perspective, if you are using touch as input then what do you want to use optikey point events for? Would you not just use native touch to perform the kind of actions that optikey can do? Perhaps you use optikey to choose a double-click, and then do a single tap in the location you want, and Optikey turns it into a double click. It seems plausible that some users would benefit from only having to do a single simple action (single tap on optikey and then off optikey) and get different outcomes. It would also make sense that some people would benefit from a magnification step. 

So I _think_ that having full functionality of point actions is important, even if it might not be necessary for all users. I also don't know how we would gracefully fall back to _not_ doing the expected thing in the case of mouse actions - I think you'd need to just disable lots of functionkeys when touch is in use, which seems undesirable. 

## options
1 - don't allow point actions when using touch

2 - hook into low level windows messages `WM_TOUCH` and `WM_POINTER` to get access to touch events everywhere. We'd also have to intercept some of the mouse click events to make sure we're not getting vanilla mouse actions when we're trying to do an Optikey mouse action. 

3 - use mouse clicks as the trigger source for touch - these can occur anywhere on the screen and should be automatically created via touch taps too. We'd have to intercept and prevent vanilla processes here too. From a brief play, I think there may be some complex interplay between touch and mouse events which makes this not necessarily a simple hack

my current instinct is to sleep on it and then probably start down the path of option 2 - but please let me know if you have any strong opinions or intel @JuliusSweetland.

**Comment by JuliusSweetland** [comment]
Hi. I think the complexity required for 2 is not worth it for the mild
benefit. I would lean towards 1, with optikey being a way of capturing
words either full screen, or to compliment usual touch screen workflows.

**Comment by kmcnaught** [comment]
how would we show this to the user? just have a list of all keys that use point actions, and disable them?

**Comment by JuliusSweetland** [comment]
Yeah that would be the more simple approach. The benefit would be that
users would easily be able to understand what is being restricted.
